### PR TITLE
feat: delta tree dispatching

### LINKED ISSUE #5200: Improve DeltaTree update routine
## Enhancement

On DeltaTree update routine (`afterNodeUpdated`), the call to `getDelta()` can be effectively translated by the compiler to SIMD memory scanning. 

image

Since compiler has already decides that `afterNodeUpdated` is too large to be inlined, we can actually dispatch this function to multiple targets. Indeed, it enjoys different codegen for `avx` targets:

image

Experiments have shown that, it can provide up to `12%` speed up at IDC environment:

```
2022-06-23T08:53:15+08:00
Running dbms/src/Storages/DeltaMerge/Benchmark/bench_deltatree
Run on (40 X 2394.97 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x20)
  L1 Instruction 32 KiB (x20)
  L2 Unified 256 KiB (x20)
  L3 Unified 25600 KiB (x2)
Load Average: 8.91, 23.44, 19.41
------------------------------------------------------------------------------------------------
Benchmark                                                      Time             CPU   Iterations
------------------------------------------------------------------------------------------------
BM_deltatree_dispatch<true>/iterations:50/threads:40     1764554 ns     59946595 ns         2000
BM_deltatree_dispatch<false>/iterations:50/threads:40    1553962 ns     55583999 ns         2000

```

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
### What problem does this PR solve?

Issue Number: close #5200

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/DeltaTree.ipp [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+DT_TEMPLATE
+template <class T>
+__attribute__((noinline, flatten)) typename DT_CLASS::InternPtr DT_CLASS::TIFLASH_DT_IMPL_NAME(T * node)
+{
+    if (!node)
+        return {};
+
+    constexpr bool is_leaf = std::is_same<Leaf, T>::value;
+
+    if (root == asNode(node) && !isLeaf(root) && node->count == 1)
```
Comment (Lloyd-Pottiger) [inline dbms/src/Storages/DeltaMerge/DeltaTree.ipp:23]: maybe  add some comments about each case? And the entire function.

Comment (SchrodingerZhu) [reply dbms/src/Storages/DeltaMerge/DeltaTree.ipp:23]: Hi, this was the original function, which is just moved to the ipp file. But yes, I guess the DeltaTree data structure itself is lack of comment.

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/DeltaMerge/DeltaTree.ipp:23]: the rest LGTM.

**Review by Lloyd-Pottiger — APPROVED** [review]

**Review by flowbehappy — APPROVED** [review]

**Comment by SchrodingerZhu** [comment]
/merge

@Lloyd-Pottiger let us see what we can add to comment in a separate PR then

**Pull request merged**